lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -4853,20 +4853,23 @@ void ConformanceChecker::ensureRequirementsAreSatisfied() {
     break;
 
   case CheckGenericArgumentsResult::RequirementFailure:
-    TypeChecker::diagnoseRequirementFailure(
-        result.getRequirementFailureInfo(), Loc, Loc,
-        proto->getDeclaredInterfaceType(), {proto->getSelfInterfaceType()},
-        QuerySubstitutionMap{substitutions}, module);
-    Conformance->setInvalid();
-    return;
   case CheckGenericArgumentsResult::SubstitutionFailure:
     // Diagnose the failure generically.
     // FIXME: Would be nice to give some more context here!
     if (!Conformance->isInvalid()) {
       proto->getASTContext().Diags.diagnose(Loc, diag::type_does_not_conform,
                                             Adoptee,
                                             Proto->getDeclaredInterfaceType());
+
+      if (result == CheckGenericArgumentsResult::RequirementFailure) {
+        TypeChecker::diagnoseRequirementFailure(
+            result.getRequirementFailureInfo(), Loc, Loc,
+            proto->getDeclaredInterfaceType(), {proto->getSelfInterfaceType()},
+            QuerySubstitutionMap{substitutions}, module);
+      }
+
       Conformance->setInvalid();
+      AlreadyComplained = true;
```

Comment:
**NOTE**

Setting `AlreadyComplained` here to avoid duplicate conformance errors caused some diagnostic regressions in value witness checking, because we skip value witness resolution when that flag is set. Ideally (as the FIXME there suggests), we should still resolve witnesses that are not dependent on failed generic requirements and type witnesses.